File: UnitTest/Zeze/Trans/TestTableNest.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Zeze.Transaction;
using Zeze.Util;

namespace UnitTest.Zeze.Trans
{
    [TestClass]
    public class TestTableNest
    {
        [TestInitialize]
        public void TestInit()
        {
            demo.App.Instance.Start();
        }

        [TestCleanup]
        public void TestCleanup()
        {
            demo.App.Instance.Stop();
        }

        [TestMethod]
        public void TestNest()
        {
            Assert.IsTrue(ResultCode.Success == demo.App.Instance.Zeze.NewProcedure(ProcTableRemove, "ProcTableRemove").CallSynchronously());
            Assert.IsTrue(ResultCode.Success == demo.App.Instance.Zeze.NewProcedure(ProcTableAdd, "ProcTableAdd").CallSynchronously());
        }

        async Task<long> ProcTableRemove()
        {
            await demo.App.Instance.demo_Module1.Table1.RemoveAsync(4321);
            return ResultCode.Success;
        }

        async Task<long> ProcTableAdd()
        {
            demo.Module1.Value v1 = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(4321);
            Assert.IsNotNull(v1);
            Assert.IsTrue(ResultCode.Success != await demo.App.Instance.Zeze.NewProcedure(
                ProcTablePutNestAndRollback, "ProcTablePutNestAndRollback").CallAsync());
            demo.Module1.Value v2 = await demo.App.Instance.demo_Module1.Table1.GetAsync(4321);
            Assert.IsNotNull(v1);
            Assert.IsTrue(v1 == v2);
            return ResultCode.Success;
        }

        async Task<long> ProcTablePutNestAndRollback()
        {
            demo.Module1.Value v = new demo.Module1.Value();
            await demo.App.Instance.demo_Module1.Table1.PutAsync(4321, v);
            return ResultCode.Unknown;
        }
    }
}

File: ZezeJava/ZezeJavaTest/src/UnitTest/Zeze/Trans/TestTableNest.java
package UnitTest.Zeze.Trans;

import demo.Module1.BValue;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import Zeze.Transaction.*;

public class TestTableNest {
	@BeforeEach
	public final void testInit() throws Exception {
		demo.App.getInstance().Start();
	}

	@AfterEach
	public final void testCleanup() throws Exception {
		//demo.App.getInstance().Stop();
	}

	@Test
	public final void testNest() throws Exception {
		Assertions.assertEquals(Procedure.Success, demo.App.getInstance().Zeze.newProcedure(TestTableNest::ProcTableRemove, "ProcTableRemove").call());
		Assertions.assertEquals(Procedure.Success, demo.App.getInstance().Zeze.newProcedure(TestTableNest::ProcTableAdd, "ProcTableAdd").call());
	}

	private static long ProcTableRemove() {
		demo.App.getInstance().demo_Module1.getTable1().remove(4321L);
		return Procedure.Success;
	}

	private static long ProcTableAdd() throws Exception {
		BValue v1 = demo.App.getInstance().demo_Module1.getTable1().getOrAdd(4321L);
		Assertions.assertNotNull(v1);
		Assertions.assertNotEquals(Procedure.Success, demo.App.getInstance().Zeze.newProcedure(TestTableNest::ProcTablePutNestAndRollback, "ProcTablePutNestAndRollback").call());
		BValue v2 = demo.App.getInstance().demo_Module1.getTable1().get(4321L);
		Assertions.assertNotNull(v1);
		Assertions.assertEquals(v1, v2);
		return Procedure.Success;
	}

	private static long ProcTablePutNestAndRollback() {
		BValue v = new BValue();
		demo.App.getInstance().demo_Module1.getTable1().put(4321L, v);
		return Procedure.Unknown;
	}
}
